### DISCUSSION TITLE
Precedence

### DISCUSSION THREAD

**Discussion by lostenderman** [lostenderman/markdown#120] [General]
@Witiko

Based on the example [341](https://spec.commonmark.org/0.30/#example-341) (see below), a `code span` should have higher precedence than `emphasis`.

```
*foo`*`
```

Looking at the [walkable_syntax][WS] however, I would assume that Emph has higher priority than Code. On the other hand, the example renders as CommonMark would expect.

```
\markdownRendererDocumentBegin
*foo\markdownRendererCodeSpan{*}\markdownRendererDocumentEnd
```

Is the precedence defined (only) in the [walkable_syntax][WS] then?

[WS]: https://github.com/Witiko/markdown/blob/e9aa96e399ef59929c2c161f85d324a5378d8db5/markdown.dtx#L1651-L1654

**Comment by Witiko** [comment]
> Based on the example [341](https://spec.commonmark.org/0.30/#example-341) (see below), a code span should have higher precedence than emphasis.

It does. However, the closing `*` will bever be found if it is consumed by `parsers.Inline` beforehand. [The `parsers.between()` pattern][1] does not do any forward-scanning. It applies `p` (here, `parsers.Inline`) repeatedly and then looks for `ender2` (here, `*` preceded by a non-space character). If `parsers.Inline` consumes the entire ``` `*` ```. there is nothing `parsers.between` can do about it and it will never even see the `*`.

 [1]: https://github.com/Witiko/markdown/blob/e9aa96e399ef59929c2c161f85d324a5378d8db5/markdown.dtx#L21499

> Is the precedence defined (only) in the [walkable_syntax](https://github.com/Witiko/markdown/blob/e9aa96e399ef59929c2c161f85d324a5378d8db5/markdown.dtx#L1651-L1654) then?

Pretty much yes. There is [the `syntax` table][1], which contains the definition of the rules in `walkable_syntax` as well as the top-level `Blocks` rule that cannot be easily represented as a list of choices. The `syntax` table is later merged with the `walkable_syntax` table after syntax extensions have been applied to `walkable_syntax`.

However, as you can see here, there is often recursion that is not captured by the precedence rules given by `walkable_syntax`. Therefore, it is often necessary to go the the definitions of the PER patterns behind the rules.

 [1]: https://github.com/Witiko/markdown/blob/91b8caf102e76649da7f5391087853548750711e/markdown.dtx#L22792-L22853
